### ISSUE TITLE
[a11y]: Tag is missing differentiator other than color

### ISSUE THREAD

**Issue by skcele** [carbon-design-system/carbon#20020]
### Package

@carbon/react

### Browser

Firefox, Chrome

### Operating System

MacOS, Windows

### Package version

v1.87.1 - unsure which version was tested but it's present in most recent version

### Automated testing tool and ruleset

Unsure, was found with AVT testing. Cited WCAG Mapping: 1.4.1 Use of Color

### Assistive technology

JAWS_Firefox

### Description

A team on IBM Cloud had an issue opened during AVT testing about using Tags and the lack of text differentiators. They brought it to the Carbon for Cloud team (us) to investigate, and we verified that even though they are using one of our components with tagging, that component uses Carbon Tags. 

[[Carbon for Cloud issue here](https://github.ibm.com/ibmcloud/cloud-pal-community/issues/1144)]

[Original issue](https://github.ibm.com/mhub/ui/issues/1989) content:

> Module: Event Streams
> Page name: Global
> 
> Page URL: https://cloud.ibm.com/services/messagehub/crn%3Av1%3Abluemix%3Apublic%3Amessagehub%3Aus-south%3Aa%2Ffdc4128df8e44504a012b1d70e7762f0%3Af2febe76-568c-4d5a-b9da-d43ce3387ab8%3A%3A?paneId=topics
> 
> Component: Tags
> 
> Issue description: Background color is used to differentiate the tag with normal text. But this is not conveyed to the screen reader user. Screen reader is announcing as a plain text, no where announced this text is a tag.
> 
> Recommended fix: Should not use color alone to convey the information /status :
> Use appropriative Text alternatives to make sure this info is available for screen reader users. or provide a region for tags so that screen reader users will understand tag region started and ended.
> 
> User impact: Screen reader users may not come to know this info.
> 
> WCAG Mapping: 1.4.1 Use of Color
> 
> Level of Compliance: A
> 
> Severity: High
> 
> Test tool: JAWS_Firefox

### WCAG 2.1 Violation

1.4.1 Use of Color

### Reproduction/example

https://stackblitz.com/edit/github-zbzqsdkf?file=src%2FApp.jsx

### Steps to reproduce

1. Use a Read-Only Tag on the page
2. Navigate using a JAWS screenreader
3. Navigate to Tag
4. Unable to differentiate that content is Tag based on what is provided

### Suggested Severity

Severity 3 = User can complete task, and/or has a workaround within the user experience of a given component.

### Code of Conduct

- [x] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [x] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

**Comment by mbgower** [comment]
@skcele I think the original issue is misconstruing [Use of Color](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html), which is entirely concerned with visual presentation. It is not related to programmatic indicators, and so the comment "this is not conveyed to the screen reader user" is unrelated to that SC.

I agree that if there is no programmatic indicator of the tag, it is a potential problem, but it would likely fall under [Info and Relationships](https://www.w3.org/WAI/WCAG22/Understanding/info-and-relationships.html), which covers the need to have a programmatic equivalent to what is visually ("presentationally") shown.

--- 
### Use of color and tags
A bit more background. Tags have always had the _potential_ to become a Use of Color issue in a couple of ways.
First, using color alone to cue there is even a tag there. Carbon's tags are designed so that there should be a 3:1 colour difference between black text and the colour used for the text of the tag; that is, a user who cannot distinguish colour should still be able to differentiate based on a 3:1 contrast. This is following a similar approach to the [technique used for non-underlined links ](https://www.w3.org/WAI/WCAG21/Techniques/general/G183).

Second, if there are multiple tags used for categorization, tags could fail Use of Color. Carbon does have guidance stressing that tag colours should reinforce textual indicators. I do not think this is the case with your use, but if for example tags are used to highlight terms used in a document, the _words_ are themselves the indicators of the term; the color just reinforces. But if the colour purple is used to tag a variety of terms related to a concept, and the color blue is used to tag a variety of terms related to a different concept, a good case could be made that those tags are relying on color for context. The (weak) counter argument would be that the highlighted terms _still_ add some additional context, so maybe it's not "color alone" indicating the material.

Note: there are ways around this second potential use of color, including by adding icons (with alts) to the tags as a secondary means of indicating a category. Carbon does have _some_ background on these kinds of considerations, but I think it would be good for the guidance to be expanded.
> Read-only tags come in several color choices and can use optional decorative icons to delineate between multiple categories.

---
### Programmatic indicators for tags
In regard to a need for programmatic indication, the context is going to matter quite a bit. If tags are used just to categorize attributes of items for sale, for example, one could make a case that the consistent positioning of the words within item tiles may be sufficient. The tag is visual _reinforcement_ and  a case could be made that it's not really necessary to surface programmatically (any more than it is other text stylings which screen readers have conventionally _not_ conveyed). In the following image the user is going to hear the item name, description, rating, price, items in store and then the tagged info. Does the visual styling of the tag convey anything meaningful? It's the same kind of debate about whether it's necessary to point out the red text in the "Save $66.50..." price info in the one item.

<img width="524" height="655" alt="detail of paint items in a commerce site showing 4 paints with 'Pick-up', 'Delivery' and 'Online Only' tags" src="https://github.com/user-attachments/assets/16594b1c-5c70-4c9d-ac0b-b94ae948f0fe" />

I do not think I would call it a failure if the above tags were just treated like normal text; I don't think it prevents a user from carrying out purchases. I'm not even sure _what_ I would be inclined to surface programmatically. Would prepending "tag" actually be helpful (e.g., "Tag: Pick-up", "Tag: Delivery")? Would it be helpful to prepend "availability"?

### Considerations for this use case
Looking at the visual in the original C4C issue, I'm not sure I would fail it either, particularly because each of the tagged items is prefaced by the term "env:". It's not a _great_ chunk of text, especially as read by a screen reader, but it seems to be used consistently with the tags. Everyone is subjected to the same crappy text abbreviation. (I can image a design where there's an icon used, and it has an alt that is not so meaningless, say "environment").

<img width="646" height="138" alt="detail of original issue image, showing the 4 tags with the following text shown: 'env:accessibili...', 'env:avt', 'env:avt1', '+4'" src="https://github.com/user-attachments/assets/f5e7d3ae-73e9-47d8-b1c5-bb42b01b7983" />

I'm a little puzzled by the design. Are these intended to be read-only tags? The "+4" tag (and the JAWS output) suggests to me it is a clickable tag, but it is visually indistinguishable from the "env "ones. That may be a contributing problem here  (and a failure of Non-text Contrast). If all these tags are actual actionable components, they need a role.  That would be a failure of 4.1.2 Name, Role, Value. 

I don't really understand the design enough to say more. So, in summary, with the info I can glean, if these are read-only tags, it does not seem a slamdunk failure. The_ impact on the user _seems_ pretty minimal, even where you want to make that case it's a failure. I'm very surprised by the sev 1 rating in the original issue. More information would help me be persuaded one way or the other. If these are _operable_ tags, I agree there seems to be a failure.

**Comment by mbgower** [comment]
BTW, I've got a query out about a possible way to semantically mark up a read-only tag. I'll let you know if I hear any good suggestions.

**Comment by mbgower** [comment]
The Carbon Accessibility Guild discussed this on our call. 
There seemed to be support for using the `mark` element for tags in situations where they are being used inline as category indicators in a read-only document. If you review this [thread in #accessibility-at-ibm](https://ibm-studios.slack.com/archives/C036P1CTN/p1620725449456100) it covers a lot of several examples of this use of tags. We don't really have anything documented on this use case in the Tags documentation.

There was some support for the idea of providing _some_ kind of programmatic indicator that a tag is not just plain text. @tombrunet suggested exploring use of `em`. @tay1orjones I think it might be worth exploring whether Carbon wanted to have all tags inside an `em` tag. Obviously CSS would be overriding the default italics style. Worth exploring (although it will not be surfaced by screen readers by default, any more than `strong` is; that's arguably not our problem).

I also realized one of the reasons I didn't find the Tags info I was looking for was because I put it into a Discussion in Carbon's github area, before I realized that this is somewhat equivalent of assigning it to Siberia. @laurenmrice [here is that discussion](https://github.com/carbon-design-system/carbon/discussions/11215), and it would be worth plumbing it for documenation.

**Comment by skcele** [comment]
Thank you for the thorough response and discussion! That thread in slack is really helpful for examples.

I like the idea of using the `em` approach for read only tags since they seem to be the primary tags that don't have other indicators of info about what they are. If it would be helpful, I can follow up with the AVT tester that raised the issue to get their input. Thank you for bringing this to the Guild!

I did some more digging into the [TagList ](https://pages.github.ibm.com/ibmcloud/cloud-pal-community/?path=/docs/connected-components-resourcetaglist--docs)component where this use case originated. The list of tags are primarily read-only but it can display tags that are clickable or dismissible (though it does not yet use the Carbon core operable and Removable tags). In most use cases, in fact almost all that I saw live in product, they are read-only. Outside of that, the `+4` tag is operable, and will open a Tooltip with the remaining Tag labels (though still not using the operable tags).

This component needs a lot of work in general in addition to migrating to the new Tags, but in terms of this issue I'll prioritize labeling the `+4` Tag appropriately and see if I can easily switch it to an Operable Tag while I'm at it.

**Comment by mbgower** [comment]
@skcele I'd expect almost any component with the name "list" in it to have a programmatic 'list' indicator, probably using a <ul> under the covers. I'm not seeing that with the TagList component you link to, and I recommend that be explored. It would also probably help users understand the context.

Jumping out of your issue question and just focusing on the TagList component, since it is not core...
I'm seeing use of `tabindex="-1"` on each of the tag items. Given they are just divs, I'm not sure what the purpose of that is. Is there a circumstance where you put focus on the tag? Otherwise, I suggest the tabindex be stripped out on non-operable components created with `div`; unless your testing shows something I'm missing, I don't think it should not affect the component, may improve it.

**Comment by laurenmrice** [comment]
Going to close this issue now.

Talked with Stephanie Cree, and she said there was a solution she was thinking about pursuing with an an em tag for read only tags to differentiate them from text, but in terms of solving the concern the AVT team raised, it seems like the work she was already planning for the TagList component should take care of it.

Also the docs have been updated to include a sentence Michael Gower recommended about Read only tags and displaying categories.
